src/digraph.rs

```diff
@@ -2265,6 +2274,159 @@ impl PyDiGraph {
+            let visit_next: Vec<NodeIndex> = nodes
+                .iter()
+                .flat_map(|n| self.graph.edges(*n))
+                .filter_map(|edge| {
+                    let target_node = edge.target();
+                    if !nodes.contains(&target_node) {
+                        Some(target_node)
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+
+            // Now, if we can reach any of `nodes`, there exists a path from `nodes`
+            // back to `nodes` of length > 1, meaning contraction is disallowed.
+            let mut dfs = Dfs::from_parts(visit_next, self.graph.visit_map());
+            while let Some(node) = dfs.next(&self.graph) {
+                if nodes.contains(&node) {
+                    // we found a path back to `nodes`
+                    return false;
+                }
+            }
+
+            true
+        };
+
+        let mut indices_to_remove: HashSet<NodeIndex> =
+            to_replace.iter().map(|&n| NodeIndex::new(n)).collect();
+
+        if check_cycle.unwrap_or(self.check_cycle)
+            && !can_contract(&indices_to_remove)
+        {
+            return Err(DAGWouldCycle::new_err(
+                "Substitution would create cycle(s)",
+            ));
+        }
+
+        // Create new node.
+        let node_index = self.graph.add_node(obj);
+
+        // Sanitize new node index from user input.
+        indices_to_remove.remove(&node_index);
+
+        // Determine edges for new node.
+        let mut incoming_edges: Vec<_> = indices_to_remove
+            .iter()
+            .flat_map(|&i| self.graph.edges_directed(i, Direction::Incoming))
+            .filter_map(|edge| {
+                let pred = edge.source();
+                if !indices_to_remove.contains(&pred) {
+                    Some((edge.source(), edge.weight().clone_ref(py)))
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        let mut outgoing_edges: Vec<_> = indices_to_remove
+            .iter()
+            .flat_map(|&i| self.graph.edges_directed(i, Direction::Outgoing))
+            .filter_map(|edge| {
+                let succ = edge.target();
+                if !indices_to_remove.contains(&succ) {
+                    Some((edge.target(), edge.weight().clone_ref(py)))
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        // Remove nodes that will be replaced.
+        for index in indices_to_remove {
+            self.graph.remove_node(index);
+        }
+
+        // If `weight_combo_fn` was specified, merge edges according
+        // to that function, even if this is a multigraph. If unspecified,
+        // defer parallel edge handling to `add_edge_no_cycle_check`.
+        if let Some(merge_fn) = weight_combo_fn {
+            macro_rules! merge_parallel_edges {
```

Comment:
Is there a reason for this to be a macro instead of a normal function? I'd actually go one step further and make it a generic function with signature like:
```rust
use std::hash::Hash;

fn merge_duplicates<K, V, F, E>(
    xs: Vec<(K, V)>,
    mut merge_fn: F,
) -> Result<Vec<(K, V)>, E>
where
    K: Hash + Eq,
    F: FnMut(&V, &V) -> Result<V, E>,
{
   ....
}
```

Reply:
I'm not strongly attached to it being a macro, but I went with this over a closure to avoid having to specify types and do error propagation (I liked the idea of the `?` operator just working in the context of where the macro is expanded). Are you proposing that a generic function with this code might be valuable as a helper method at the class level?

Reply:
Yeah exactly. We can place it in `src/lib.rs`  and re-use it in the sister method for `PyGraph`.

Reply:
Done in 02f491a. Out of curiosity, is there a reason / best practice for making `F` be `FnMut`, rather than `Fn`?

Reply:
It's just that `Fn` is a subtrait of `FnMut` so anything that implements `Fn` automatically implements `FnMut`. See https://doc.rust-lang.org/std/ops/trait.Fn.html